onnxruntime/core/providers/coreml/model/model.mm

```diff
@@ -498,88 +294,232 @@ Status GetMLMultiArrayCopyInfo(const MLMultiArray* _Nonnull array,
 }
 
 // Internal Execution class
-// This class will bridge Model (c++) with CoreMLExecution (objective c++)
+// This class is part of the model class and handles the calls into CoreML
 class Execution {
  public:
   Execution(const std::string& path, const logging::Logger& logger, uint32_t coreml_flags);
-  ~Execution() {};
+  ~Execution();
 
   Status LoadModel();
   Status Predict(const std::unordered_map<std::string, OnnxTensorData>& inputs,
                  const std::unordered_map<std::string, OnnxTensorInfo>& outputs,
                  const GetOutputTensorMutableRawDataFn& get_output_tensor_mutable_raw_data_fn);
 
  private:
-  bool model_loaded{false};
-  CoreMLExecution* execution_;
+  void cleanup();
+  bool model_loaded_{false};
+  NSString* coreml_model_path_{nil};
+  NSString* compiled_model_path_{nil};
+  const logging::Logger& logger_;
+  uint32_t coreml_flags_{0};
+  MLModel* model_{nil};
 };
 
-Execution::Execution(const std::string& path, const logging::Logger& logger, uint32_t coreml_flags) {
+Execution::Execution(const std::string& path, const logging::Logger& logger, uint32_t coreml_flags)
+    : logger_(logger), coreml_flags_(coreml_flags) {
   @autoreleasepool {
-    execution_ = [[CoreMLExecution alloc] initWithPath:path
-                                                logger:logger
-                                          coreml_flags:coreml_flags];
+    coreml_model_path_ = util::Utf8StringToNSString(path.c_str());
+  }
+}
+
+Execution::~Execution() { cleanup(); }
```

Comment:
to be safe, perhaps we should put any non-trivial Objective-C code into an autorelease block. even if it may not be required now, it's future proof if some code that produces autoreleased objects is added later.
```suggestion
Execution::~Execution() {
  @autoreleasepool {
    cleanup();
  }
}
```